config/v3_0/types/disk.go

```diff
@@ -16,72 +16,68 @@ package types
 
 import (
 	"github.com/coreos/ignition/v2/config/shared/errors"
-	"github.com/coreos/ignition/v2/config/validate/report"
+
+	"github.com/ajeddeloh/vcontext/path"
+	"github.com/ajeddeloh/vcontext/report"
 )
 
 func (d Disk) Key() string {
 	return d.Device
 }
 
-func (n Disk) Validate() report.Report {
-	return report.Report{}
-}
-
-func (n Disk) ValidateDevice() (r report.Report) {
+func (n Disk) Validate(c path.ContextPath) (r report.Report) {
 	if len(n.Device) == 0 {
-		r.AddOnError(errors.ErrDiskDeviceRequired)
+		r.AddOnError(c.Append("device"), errors.ErrDiskDeviceRequired)
 		return
 	}
-	r.AddOnError(validatePath(n.Device))
-	return
-}
+	r.AddOnError(c.Append("device"), validatePath(n.Device))
 
-func (n Disk) ValidatePartitions() (r report.Report) {
-	if n.partitionNumbersCollide() {
-		r.AddOnError(errors.ErrPartitionNumbersCollide)
+	if collides, p := n.partitionNumbersCollide(); collides {
+		r.AddOnError(c.Append("partitions", p), errors.ErrPartitionNumbersCollide)
 	}
-	if n.partitionsOverlap() {
-		r.AddOnError(errors.ErrPartitionsOverlap)
+	if overlaps, p := n.partitionsOverlap(); overlaps {
+		r.AddOnError(c.Append("partitions", p), errors.ErrPartitionsOverlap)
 	}
 	if n.partitionsMixZeroesAndNonexistence() {
-		r.AddOnError(errors.ErrZeroesWithShouldNotExist)
+		r.AddOnError(c.Append("partitions"), errors.ErrZeroesWithShouldNotExist)
 	}
-	if n.partitionLabelsCollide() {
-		r.AddOnError(errors.ErrDuplicateLabels)
+	if collides, p := n.partitionLabelsCollide(); collides {
+		r.AddOnError(c.Append("partitions", p), errors.ErrDuplicateLabels)
 	}
 	return
 }
 
-// partitionNumbersCollide returns true if partition numbers in n.Partitions are not unique.
-func (n Disk) partitionNumbersCollide() bool {
-	m := map[int][]Partition{}
-	for _, p := range n.Partitions {
+// partitionNumbersCollide returns true if partition numbers in n.Partitions are not unique. It also returns the
+// index of the colliding partition
+func (n Disk) partitionNumbersCollide() (bool, int) {
+	m := map[int][]int{} // from partition number to index into array
+	for i, p := range n.Partitions {
 		if p.Number != 0 {
 			// a number of 0 means next available number, multiple devices can specify this
-			m[p.Number] = append(m[p.Number], p)
+			m[p.Number] = append(m[p.Number], i)
 		}
 	}
 	for _, n := range m {
 		if len(n) > 1 {
 			// TODO(vc): return information describing the collision for logging
-			return true
+			return true, n[1]
 		}
 	}
-	return false
+	return false, 0
```

Comment:
Even though 0 isn't a valid return value thoughts on using a negative number when returning `false` to reinforce that the partition was valid in these validation functions?

Reply:
I think it's pretty standard to return the default go zero value for unused returns. Like when return `types.Config{}`

Reply:
Fair, I was thinking in the context of functions like [`sort.Search`](https://golang.org/pkg/sort/#Search) where it returns `-1` when a value is not found.